plugins/rest/rest_auth.go

```diff
@@ -113,33 +138,116 @@ type tokenEndpointResponse struct {
 // oauth2ClientCredentialsAuthPlugin represents authentication via a bearer token in the HTTP Authorization header
 // obtained through the OAuth2 client credentials flow
 type oauth2ClientCredentialsAuthPlugin struct {
-	TokenURL     string    `json:"token_url"`
-	ClientID     string    `json:"client_id"`
-	ClientSecret string    `json:"client_secret"`
-	Scopes       *[]string `json:"scopes,omitempty"`
-
-	tokenCache    *oauth2Token
-	tlsSkipVerify bool
+	GrantType    string                 `json:"grant_type"`
+	TokenURL     string                 `json:"token_url"`
+	ClientID     string                 `json:"client_id"`
+	ClientSecret string                 `json:"client_secret"`
+	SigningKeyID string                 `json:"signing_key"`
+	Claims       map[string]interface{} `json:"additional_claims"`
+	IncludeJti   bool                   `json:"include_jti_claim"`
+	Scopes       *[]string              `json:"scopes,omitempty"`
+
+	signingKey       *keys.KeyConfig
+	signingKeyParsed interface{}
+	tokenCache       *oauth2Token
+	tlsSkipVerify    bool
 }
 
 type oauth2Token struct {
 	Token     string
 	ExpiresAt time.Time
 }
 
+func (ap *oauth2ClientCredentialsAuthPlugin) createAuthJwt(claims map[string]interface{}, signingKey interface{}) (*string, error) {
+	now := time.Now()
+	baseClaims := map[string]interface{}{
+		"iat": now.Unix(),
+		"exp": now.Add(1 * time.Hour).Unix(),
+	}
+	for k, v := range baseClaims {
+		claims[k] = v
+	}
+
+	if ap.Scopes != nil && len(*ap.Scopes) > 0 {
+		claims["scope"] = strings.Join(*ap.Scopes, " ")
+	}
```

Comment:
Hmm. I had been wondering why we had `*[]string` and not `[]string`, and then I thought that we'd perhaps want to differentiate _empty slice_ from _not set_; but it looks like we don't do that here.

So... why not `[]string`?

(Also, sideline, we could still differentiate `[]string(nil)` (unset) from `[]string{}` (set but empty), I think.)

Reply:
I'll look into it, thanks.